docs/Style.thy

```diff
@@ -0,0 +1,263 @@
+
+lemma disj_conj_distribR:
+  "(A \<and> B \<or> C) = ((A \<or> C) \<and> (B \<or> C))"
+  apply (rule iffI)
+   apply (rule conjI)
+    apply fast
+   apply fast
+  apply auto
+  done
+
+text \<open>
+The statement of Hoare triple lemmas are generally indented in the following way.
+\<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace>precondition_one and precondition_two and
+    precondition three\<rbrace>
+   my_function param_a param_b
+   \<lbrace>post_condition\<rbrace>"
+  oops
+
+text \<open> or \<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace>precondition_one and precondition_two
+    and precondition three\<rbrace>
+   my_function param_a param_b
+   \<lbrace>post_condition\<rbrace>"
+  oops
+
+text \<open>
+Definitions, record and datatypes are generally indented as follows and generally use \<equiv> rather than
+=. Short datatypes can be on a single line.
+
+Note: Upstream isabelle use tends to use = rather than \<equiv>. So we don't enforce a strict rule in this
+case.
+\<close>
+
+definition word_bits :: nat where
+  "word_bits \<equiv> 32"
+
+definition long_defn_type ::
+  "'this => 'is => 'a => 'long => 'type => 'that => 'does =>
+   'not => 'fit => 'in => 'one => 'line"
+  where
+  "long_defn_type \<equiv> undefined"
+
+fun foo_fun :: "nat \<Rightarrow> nat" where
+    "foo_fun 0 = 0"
+  | "foo_fun (Suc n) = n"
+
+record cdl_cnode =
+  cdl_cnode_caps :: type_foo
+  cdl_cnode_size_bits :: type_bar
+
+datatype cdl_object =
+    Endpoint
+  | Tcb type_foo
+  | CNode cdl_cnode
+  | Untyped
+
+datatype cdl_arch = IA32 | ARM11
+
+text \<open>
+There are tools to automatically indent proofs in jEdit. In terms of rules:
+  * definition and fun should be on the same line as the name.
+  * Everything after the first line of a definition or fun statement should be indented.
+  * Type definitions (record, type_synonym, datatype) have the defined type on the same line, the
+    rest only if it fits.
+  * Avoid long hanging indentation in data types, case expressions, etc. That is:
+
+Wrong:
+
+@{term\<open>
+case xs_blah of [] \<Rightarrow> foo
+              | y#ys \<Rightarrow> baz
+\<close>}
+
+Right:
+
+@{term\<open>
+case xs_blah of
+    [] \<Rightarrow> foo
+  | y#ys \<Rightarrow> baz
+\<close>}
+\<close>
+
+text \<open>
+  If a definition body causes an overflow of the 100 char line limit, use one
+  of the following two patterns.
+\<close>
+
+definition hd_opt :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt \<equiv> case_list
+              None
+              (\<lambda>h t. Some h)"
+
+definition hd_opt4 :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt4 \<equiv>
+   case_list None (\<lambda>h t. Some h)"
+
+text \<open>
+  If an apply line causes an overflow of the 100 char line limit, use one
+  of the following two patterns: aligning on ":" or aligning on left.
+\<close>
+
+lemma test_lemma0:
+  "\<lbrakk>hd_opt l \<noteq> None; hd_opt (hd l) = Some x\<rbrakk> \<Longrightarrow>
+    hd_opt (concat l) = Some x"
```

Comment:
Small detail: in this example is hd_opt indented one too many here? (same in test_lemma1); it seems it's because of the `\<lbrakk>` but I'm not familiar of doing things this way.

Reply:
> > > > One more comment: this should probably be included in a session somewhere. Is there a docs session?
> > > 
> > > 
> > > There is no `ROOT` in `l4v/docs/`, and no `docs` in `l4v/ROOTS`. We can make it though.
> > 
> > 
> > I just added a rather quickly assembled thing that appears to add this to the regression, according to some very superficial testing. Might be overkill; what do people think?
> 
> I think it's great!

Looks good to me!